packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -5238,16 +5238,33 @@ class InheritedElement extends ProxyElement {
   InheritedElement(InheritedWidget super.widget);
 
   final Map<Element, Object?> _dependents = HashMap<Element, Object?>();
+  final Map<Type, InheritedElement> _lookupTable = HashMap<Type, InheritedElement>();
+
+  @pragma('vm:prefer-inline')
+  void _updateLookupTable() {
+    if (_lookupTable.isNotEmpty) {
+      return;
+    }
+    _lookupTable[widget.runtimeType] = this;
+    InheritedElement? ancestor = _inheritedAncestor;
+    while (ancestor != null) {
+      // Reading out the ancestor's lookup table is slower than just running
+      // up the tree based on local experiments.
+      _lookupTable.putIfAbsent(ancestor.widget.runtimeType, () => ancestor!);
```

Comment:
how do you avoid overrwriting types that are provided more than once?

Reply:
putIfAbsent means I'm taking the first one I find as I traverse up the tree.

Reply:
oh I see!

`_lookupTable[ancestor.widget.runtimeType] ??= ancestor!`

Will be marginally faster, especially on web

Reply:
Going with it.